mods/page_revisions.js

```diff
@@ -351,66 +483,68 @@ PRS.prototype.getTitleRevision = function(restbase, req) {
+        revisionRequest = P.props({
+            revisionInfo: restbase.get({
+                uri: this.tableURI(rp.domain),
+                body: {
+                    table: this.tableName,
+                    attributes: {
+                        title: title,
+                        rev: parseInt(rp.revision)
+                    },
+                    limit: 1
+                }
+            })
+            .catch(function(e) {
+                if (e.status !== 404) {
+                    throw e;
+                }
+                return self.fetchAndStoreMWRevision(restbase, req);
+            }),
+            pageData: pageDataRequest
         });
     } else if (!rp.revision) {
         if (req.headers && /no-cache/.test(req.headers['cache-control'])) {
-            revisionRequest = self.fetchAndStoreMWRevision(restbase, req)
+            revisionRequest = P.props({
+                revisionInfo: self.fetchAndStoreMWRevision(restbase, req),
+                pageData: pageDataRequest
+            })
+            .then(function(res) {
+                return self._checkPageUndeleted(restbase, req, res);
+            })
             .catch(function(e) {
                 if (e.status !== 404) {
                     throw e;
                 }
-                return getLatestTitleRevision()
-                // In case 404 is returned by MW api, the page is deleted
-                .then(function(result) {
-                    result = result.body.items[0];
-                    result.tid = uuid.now().toString();
-                    result.restrictions = result.restrictions || [];
-                    result.restrictions.push('page_deleted');
-                    return restbase.put({
-                        uri: self.tableURI(rp.domain),
-                        body: {
-                            table: self.tableName,
-                            attributes: result
-                        }
-                    })
-                    .then(function() {
-                        throw e;
-                    });
+                // Page data request never returns 404, so it's from MW API => page was deleted
+                return self._signalPageDeleted(restbase, req)
+                .then(function() {
+                    throw e;
                 });
             });
         } else {
-            revisionRequest = getLatestTitleRevision()
-            .catch(function(e) {
-                if (e.status !== 404) {
-                    throw e;
-                }
-                return self.fetchAndStoreMWRevision(restbase, req);
+            revisionRequest = P.props({
+                revisionInfo: getLatestTitleRevision()
+                .catch(function(e) {
+                    if (e.status !== 404) {
+                        throw e;
+                    }
+                    return self.fetchAndStoreMWRevision(restbase, req);
+                }),
+                pageData: pageDataRequest
```

Comment:
Actually, we did call `checkRevReturn` for the latest revision before the delete detection was added too: https://github.com/wikimedia/restbase/pull/269/files

Also, we still have to check for `page_deleted` restriction, because for older content the page table doesn't contain that information.